linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -42,13 +42,17 @@ final class Emitter(config: Emitter.Config) {
 
   private val uncachedKnowledge = new knowledgeGuardian.KnowledgeAccessor {}
 
+  private val nameCompressor: Option[NameCompressor] =
+    if (minify) Some(new NameCompressor(config))
+    else None
```

Comment:
Shouldn't this be part of the state as well?

Reply:
In theory, it probably should. But to create the state we need the `lastMentionedDangerousGlobalRefs`, and to compute those, we need the `nameCompressor`. So that would be tricky to manipulate. It seems easier to manage if it is outside.

Reply:
Right. So let's table that discussion until we're settled on how to collect the names we want to minify.

Reply:
Going along with this: The duplication of the logic in the `NameCompressor` and the rest of the `Emitter` still makes me quite uneasy (even with the usage validation, IMHO it is not very easy to verify that the counting logic is correct).

If we use a return type, we could return different implementations that track or do not track the exact count.

I realize you were concern about not being able to approximate, but IIUC the only place where we do this right now is for Apply to hijacked classes. But this seems to be easy to do correctly:

- `isMaybeHijackedClass` is already factored in a helper.
- We can use the `uncachedGlobalKnowledge` to determine this in the `NameCompressor`.
- Updating the global knowledge needs to be moved to `emit` from `emitInternal` but that does not seem to be an issue.